File: BlockBattleBot/Piece.cs
using System.Collections.Generic;

namespace BlockBattleBot
{
    public class Piece
    {
        public PieceType PieceType { get; set; }

        public int Rotations { get; set; }

        #region Piece Cells

        private static Dictionary<PieceType, byte[][,]> cells = new Dictionary<PieceType, byte[][,]>() {
            { PieceType.I, new byte[][,] { 
                new byte[,] { 
                 { 0, 0, 0, 0},
                 { 1, 1, 1, 1},
                 { 0, 0, 0, 0},
                 { 0, 0, 0, 0} },

                new byte[,] {
                 { 0, 0, 1, 0},
                 { 0, 0, 1, 0},
                 { 0, 0, 1, 0},
                 { 0, 0, 1, 0} },

                new byte[,] {
                 { 0, 0, 0, 0},
                 { 0, 0, 0, 0},
                 { 1, 1, 1, 1},
                 { 0, 0, 0, 0} },

                new byte[,] {
                 { 0, 1, 0, 0},
                 { 0, 1, 0, 0},
                 { 0, 1, 0, 0},
                 { 0, 1, 0, 0} }
            }},

            { PieceType.J, new byte[][,] {
                new byte[,] {
                  { 1, 0, 0 },
                  { 1, 1, 1 },
                  { 0, 0, 0 } },

                new byte[,] {
                  { 0, 1, 1 },
                  { 0, 1, 0 },
                  { 0, 1, 0 } },

                new byte[,] {
                  { 0, 0, 0 },
                  { 1, 1, 1 },
                  { 0, 0, 1 } },

                new byte[,] {
                  { 0, 1, 0 },
                  { 0, 1, 0 },
                  { 1, 1, 0 } }
            }},

            { PieceType.L, new byte[][,] {
                new byte[,] {
                  { 0, 0, 1 },
                  { 1, 1, 1 },
                  { 0, 0, 0 } },

                new byte[,] {
                  { 0, 1, 0 },
                  { 0, 1, 0 },
                  { 0, 1, 1 } },

                new byte[,] {
                  { 1, 1, 1 },
                  { 1, 0, 0 },
                  { 0, 0, 0 } },

                new byte[,] {
                  { 1, 1, 0 },
                  { 0, 1, 0 },
                  { 0, 1, 0 } }
            }},

            { PieceType.O, new byte[][,] {
                new byte[,] {
                  { 1, 1 },
                  { 1, 1 } }
            }},

            { PieceType.S, new byte[][,] {
                new byte[,] {
                  { 0, 1, 1 },
                  { 1, 1, 0 },
                  { 0, 0, 0 } },

                new byte[,] {
                  { 1, 0, 0 },
                  { 1, 1, 1 },
                  { 0, 0, 1 } }
            }},

            { PieceType.T, new byte[][,] {
                new byte[,] {
                  { 0, 1, 0 },
                  { 1, 1, 1 },
                  { 0, 0, 0 } },

                new byte[,] {
                  { 0, 1, 0 },
                  { 0, 1, 1 },
                  { 0, 1, 0 } },

                new byte[,] {
                  { 0, 0, 0 },
                  { 1, 1, 1 },
                  { 0, 1, 0 } },

                new byte[,] {
                  { 0, 1, 0 },
                  { 1, 1, 0 },
                  { 0, 1, 0 } }
            }},

            { PieceType.Z, new byte[][,] {
                new byte[,] {
                  { 1, 1, 0 },
                  { 0, 1, 1 },
                  { 0, 0, 0 } },

                new byte[,] {
                  { 0, 0, 1 },
                  { 0, 1, 1 },
                  { 0, 1, 0 } }
            }}
        };

        #endregion

        public Piece(PieceType type, int rotations = 0)
        {
            PieceType = type;
            Rotations = rotations;
        }

        public int MaxRotations
        {
            get
            {
                return cells[PieceType].GetLength(0);
            }
        }

        public byte[,] Cells
        {
            get
            {
                return cells[PieceType][Rotations];
            }
        }
    }
}

File: src/Piece.ts
import { PieceType } from "./PieceType";

interface PieceCells {
	[type: string]: number[][][]
}

let cells: PieceCells = {
	"I": 
	[
		[
			[ 0, 0, 0, 0 ],
            [ 1, 1, 1, 1 ],
            [ 0, 0, 0, 0 ],
            [ 0, 0, 0, 0 ]
		], [
			[ 0, 0, 1, 0 ],
			[ 0, 0, 1, 0 ],
			[ 0, 0, 1, 0 ],
			[ 0, 0, 1, 0 ]
		], [
			[ 0, 0, 0, 0 ],
			[ 0, 0, 0, 0 ],
			[ 1, 1, 1, 1 ],
			[ 0, 0, 0, 0 ]
		], [
			[ 0, 1, 0, 0 ],
			[ 0, 1, 0, 0 ],
			[ 0, 1, 0, 0 ],
			[ 0, 1, 0, 0 ] 
		]
	],
	
	"J":
	[
		[
			[ 1, 0, 0 ],
			[ 1, 1, 1 ],
			[ 0, 0, 0 ]
		], [
			[ 0, 1, 1 ],
			[ 0, 1, 0 ],
			[ 0, 1, 0 ]
		], [
			[ 0, 0, 0 ],
			[ 1, 1, 1 ],
			[ 0, 0, 1 ]
		], [
			[ 0, 1, 0 ],
			[ 0, 1, 0 ],
			[ 1, 1, 0 ] 
		]
	],
	
	"L":
	[
		[
			[ 0, 0, 1 ],
			[ 1, 1, 1 ],
			[ 0, 0, 0 ]
		], [
			[ 0, 1, 0 ],
			[ 0, 1, 0 ],
			[ 0, 1, 1 ]
		], [
			[ 1, 1, 1 ],
			[ 1, 0, 0 ],
			[ 0, 0, 0 ]
		], [
			[ 1, 1, 0 ],
			[ 0, 1, 0 ],
			[ 0, 1, 0 ]
		]
	],
	
	"O":
	[
		[
			[ 1, 1 ],
			[ 1, 1 ]
		]
	],
	
	"S":
	[
		[
			[ 0, 1, 1 ],
			[ 1, 1, 0 ],
			[ 0, 0, 0 ]
		], [
			[ 1, 0, 0 ],
			[ 1, 1, 1 ],
			[ 0, 0, 1 ] 
		]
	],
	
	"T":
	[
		[
			[ 0, 1, 0 ],
			[ 1, 1, 1 ],
			[ 0, 0, 0 ]
		], [
			[ 0, 1, 0 ],
			[ 0, 1, 1 ],
			[ 0, 1, 0 ]
		], [
			[ 0, 0, 0 ],
			[ 1, 1, 1 ],
			[ 0, 1, 0 ]
		], [
			[ 0, 1, 0 ],
			[ 1, 1, 0 ],
			[ 0, 1, 0 ]
		]
	],
	
	"Z":
	[
		[
			[ 1, 1, 0 ],
			[ 0, 1, 1 ],
			[ 0, 0, 0 ]
		], [
			[ 0, 0, 1 ],
			[ 0, 1, 1 ],
			[ 0, 1, 0 ]
		]
	]
};

export class Piece {
	public pieceType: PieceType;
	private _rotations: number;
	
	constructor(type: PieceType, rotations:number = 0) {
		this.pieceType = type;
		this.rotations = rotations;
	}
	
	public get maxRotations(): number {
		let type: string = PieceType[this.pieceType];
		
		return cells[type].length;
	}
	
	public set rotations(rotations: number) {
		if (rotations >= this.maxRotations) {
			rotations = rotations % this.maxRotations;
		}
		
		this._rotations = rotations;
	}
	
	public get cells(): number[][] {
		let type: string = PieceType[this.pieceType];
		
		return cells[type][this._rotations];
	}
}